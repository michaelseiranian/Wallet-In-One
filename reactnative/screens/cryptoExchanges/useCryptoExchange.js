import { useState } from "react";
import * as SecureStore from "expo-secure-store";
import { BACKEND_URL } from "@env"
import { api_url } from '../../authentication';


export default function useCryptoExchange() {
  const [exchanges, setExchanges] = useState([]);
  // Fetch the crypto exchange accounts
  const fetchExchanges = async () => {
    await fetch(`${api_url}/crypto-exchanges`, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Token ${await SecureStore.getItemAsync("token")}`,
        },
      })
        .then((res) => res.json())
        .then((res) => setExchanges(res))
        .catch((err) => console.log(err));
  }
  // Remove exchange backend call
  const removeExchange = async (id) => {
    await fetch(`${api_url}/crypto-exchanges`, {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Token ${await SecureStore.getItemAsync("token")}`,
      },
      body: JSON.stringify({
        id: id
      }),
    })
      .then(() => setExchanges(exchanges.filter(exchange => exchange.id !== id)))
      .catch((err) => console.log(err));

    }

  return { exchanges, setExchanges, fetchExchanges, removeExchange };
}