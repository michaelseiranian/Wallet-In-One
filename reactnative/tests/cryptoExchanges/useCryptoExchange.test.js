import { renderHook } from '@testing-library/react-hooks';
import useCryptoExchange from "../../screens/cryptoExchanges/useCryptoExchange";
import expect from "expect";
import {describe, test} from "@jest/globals";
import {act} from "@testing-library/react-native";


describe('useCryptoExchange', () => {
    test('should match snapshot', () => {
        const { result } = renderHook(() => useCryptoExchange());

        expect(result.current).toMatchSnapshot();
    });

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('fetchExchanges works correctly', async () => {
        const mockFetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve([{ id: 1, name: 'Exchange 1' }]),
            })
        );
        global.fetch = mockFetch;

        const { result, waitForNextUpdate } = renderHook(() => useCryptoExchange());
        expect(result.current.exchanges).toEqual([]);
    await act( async () => {
        await result.current.fetchExchanges();

        expect(mockFetch).toHaveBeenCalledWith(expect.stringContaining(`/crypto-exchanges`), {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                Authorization: `Token undefined`,
            },
        });
        expect(result.current.exchanges).toEqual([{ id: 1, name: 'Exchange 1' }]);
    });
    });

    test('fetchExchange rejects expectedly', async () => {
        const mockFetch = jest.fn(() => {
            return Promise.reject();
        })

        global.fetch = mockFetch;
        const {result} = renderHook(() => useCryptoExchange());

        await act(async () => {
            result.current.setExchanges([{id: 1, name: 'Exchange 1'}]);

            await result.current.fetchExchanges();
        });
        expect(mockFetch).toHaveBeenCalledWith(expect.stringContaining(`/crypto-exchanges`), {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                Authorization: `Token undefined`,
            }
        });
    });

    test('removeExchange works correctly', async () => {
            const mockFetch = jest.fn(() =>
                Promise.resolve({
                    json: () => Promise.resolve([{id: 1, name: 'Exchange 1'}, {id: 2, name: 'Exchange 2'}]),
                })
            );
            global.fetch = mockFetch;
            const {result} = renderHook(() => useCryptoExchange());


        await act(async () => {

            await result.current.setExchanges([{id: 1, name: 'Exchange 1'}, {id: 2, name: 'Exchange 2'}]);
            await result.current.removeExchange(1);

            expect(mockFetch).toHaveBeenCalledWith(expect.stringContaining(`/crypto-exchanges`), {
                method: 'DELETE',
                headers: {
                    'Content-Type': 'application/json',
                    Authorization: `Token undefined`,
                },
                body: JSON.stringify({
                    id: 1,
                }),
            });
            expect(result.current.exchanges).toEqual([{id: 2, name: 'Exchange 2'}]);
        });

    });

    test('removeExchange rejects expectedly', async () => {
        const mockFetch = jest.fn(() => {
            return Promise.reject();
        })

        global.fetch = mockFetch;
        const {result} = renderHook(() => useCryptoExchange());

        await act(async () => {
            result.current.setExchanges([{use: 1, name: 'Exchange 1'}]);

            await result.current.removeExchange(1);
        });
            expect(mockFetch).toHaveBeenCalledWith(expect.stringContaining(`/crypto-exchanges`), {
                method: 'DELETE',
                headers: {
                    'Content-Type': 'application/json',
                    Authorization: `Token undefined`,
                },
                body: JSON.stringify({
                    id: 1,
                }),
            });
    });


});