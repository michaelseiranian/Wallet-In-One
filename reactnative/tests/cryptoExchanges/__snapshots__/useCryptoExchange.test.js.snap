// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCryptoExchange should match snapshot 1`] = `
{
  "exchanges": [],
  "fetchExchanges": [Function],
  "removeExchange": [Function],
  "setExchanges": [Function],
}
`;
